src/main/java/duke/task/TaskManager.java

```diff
@@ -0,0 +1,264 @@
+import java.util.ArrayList;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import duke.DukeException;
+import duke.parser.DateHandler;
+
+/**
+ * Manager class to manage and keep track of all the actions being performed on the task.
+ */
+public class TaskManager {
+    private static final String listingResponse = "Here are the tasks in your list:";
+    //Strings for marking and unmarking
+    private static final String RESPONSE_MARK = "Nice! I've marked this task as done:";
+    private static final String RESPONSE_UMARK = "OK, I've marked this task as not done yet:";
+    private static final String RESPONSE_REMOVE = "Noted. I've removed this task";
+    //String and variables for task
+    private static final String RESPONSE_ADD = "Got it. I've added this task:";
+    private static final String RESPONSE_FIND = "Here are the matching tasks in your list";
+    private final ArrayList<Task> items;
+    private boolean hasChanged = false;
+
+
+    /**
+     * Creates a manager.
+     */
+    public TaskManager() {
+        this.items = new ArrayList<>();
+    }
+
+    /**
+     * Adds task to the current manager list.
+     *
+     * @param item A task to add.
+     */
+    public void addItem(Task item) {
+        items.add(item);
+    }
+
+    /**
+     * Adds a Task based on the possible Actions and instruction to the current manage
+     *
+     * @param options     A valid Action to perform.
+     * @param instruction A string value of the commands.
+     * @return An ArrayList of string of the values to output to the Ui.
+     * @throws DukeException Invalid processing of the items.
+     */
+    public String[] addTask(Actions options, String instruction) throws DukeException {
+        Task item;
+        String description;
+        String by;
+        String from;
+        Pattern deadlineFormat = Pattern.compile("(?<description>.+)\\s?((?i)/by)(?<by>.+)");
+        Pattern eventFormat = Pattern.compile("(?<description>.+)\\s?((?i)/from)(?<from>.+)((?i)/to)(?<by>.+)");
+        switch (options) {
+        case TODO:
+            if (instruction.isBlank()) {
+                throw new DukeException("description");
+            }
+            item = new Todo(instruction);
+            break;
+        case DEADLINE:
+            Matcher deadlineMatch = deadlineFormat.matcher(instruction);
+            if (!deadlineMatch.find()) {
+                throw new DukeException("GIGABOOOM");
+            }
+            description = deadlineMatch.group("description");
+            by = deadlineMatch.group("by").trim();
+            if (description.isBlank()) {
+                throw new DukeException("description");
+            } else if (by.isBlank()) {
+                throw new DukeException("by");
+            }
+            Optional<LocalDate> testDate = DateHandler.checkDate(by);
+            item = testDate
+                    .map(localDate -> new Deadline(description,
```

Comment:
I like how your code is written with lots of functional programming concepts in mind. I did not think of using Optional, but looking at this, it's definitely something I would want to consider now.